### PR TITLE
Add compiler support for offloading matmul ops to custom dispatch plugin.

### PR DESCRIPTION
**Pull request by monorimet**
Add lit test and fix build.

Fixes to LowerToAccelUKernelPass

Tweaks to LowerToAccelUKernelsPass.

Add AccelMatmulExpert pass pipeline

Apply clang-format to new C++ files. (#3)

- Apply clangformat.

Use parameter struct calling convention

Tweaks to KernelDispatch and AccelMatmulExpert pipeline

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Probably needs a rebase.

**Comment by monorimet** [comment]
Auto-rebase 🫠

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp [hunk 1]
```diff
@@ -1159,6 +1159,18 @@ static LogicalResult setRootConfig(func::FuncOp entryPointFn,
       DispatchLoweringPassPipeline::Mmt4dTilingExpert);
 }
 
+/// Sets the lowering configuration for dispatch region for linalg.matmul root
+/// op
+static LogicalResult setRootConfig(func::FuncOp entryPointFn,
+                                   linalg::MatmulOp matmulOp) {
+  assert(!getLoweringConfig(matmulOp) && "expected lowering_config is not set");
+  SmallVector<int64_t> tileSizes;
+  tileSizes.push_back(1);
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:1285]: Why do you need `{1}` here? Are you actually tiling it by 1?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerToAccelUKernels.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+  Location loc = op.getLoc();
+  Value m = rewriter.create<tensor::DimOp>(loc, lhs, 0);
+  Value n = rewriter.create<tensor::DimOp>(loc, rhs, 0);
+  Value k = rewriter.create<tensor::DimOp>(loc, rhs, 1);
+
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(op);
+  auto fn = getFnNameAndDefAttrs("accel_matmul_f32", rewriter, targetAttr);
+  auto genericMicroKernelOp = rewriter.create<IREE::Codegen::UKernelGenericOp>(
+      loc, outType, fn.name, ValueRange{lhs, rhs}, out, ValueRange{m, n, k},
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerToAccelUKernels.cpp:93]: Are you sure the plugin is taking `{m, n, k}` last I saw it was `{m, k, n}`. Just need to be consistent.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerToAccelUKernels.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+/// Returns the function name and attributes to use for a ukernel with given
+/// `ukernelName` on the target described by `targetAttr`.
+static FnNameAndDefAttrs
+getFnNameAndDefAttrs(const char *ukernelName, RewriterBase &rewriter,
+                     IREE::HAL::ExecutableTargetAttr targetAttr) {
+  FnNameAndDefAttrs result;
+  result.name = ukernelName;
+  result.defAttrs.emplace_back(
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerToAccelUKernels.cpp:63]: Only add these if the plugin needs it. If it doesnt, just lesser things to deal with.

**Review by MaheshRavishankar — COMMENTED** [review]
This looks mostly fine. Have a few comments (also marked draft). Would be good to add an e2e test with the flag enabled to check for results on a matmul calling the plugin.

**Pull request closed without merging**